balancer/rls/internal/cache.go

```diff
@@ -0,0 +1,403 @@
+			dc.logger.Warningf("cachekey %+v is too recent to be evicted. Stopping cache resizing for now", key)
+			break
+		}
+
+		// Stop the backoff timer before evicting the entry.
+		if entry.backoffState != nil && entry.backoffState.timer != nil {
+			if entry.backoffState.timer.Stop() {
+				entry.backoffState.timer = nil
+				backoffCancelled = true
+			}
+		}
+		dc.deleteAndcleanup(key, entry)
+	}
+	dc.targetSize = size
+	return backoffCancelled
+}
+
+// evictExpiredEntries sweeps through the cache and deletes expired entries. An
+// expired entry is one for which both the `expiryTime` and `backoffExpiryTime`
+// fields are in the past.
+//
+// The return value indicates if any expired entries were evicted.
+//
+// The LB policy invokes this method periodically to purge expired entries.
+func (dc *dataCache) evictExpiredEntries() bool {
+	if dc.shutdown.HasFired() {
+		return false
+	}
+
+	evicted := false
+	dc.keys.iterateAndRun(func(key cacheKey) {
+		entry, ok := dc.entries[key]
+		if !ok {
+			// This should never happen.
+			dc.logger.Errorf("cacheKey %+v not found in the cache while attempting to perform periodic cleanup of expired entries", key)
+			return
+		}
+
+		// Only evict entries for which both the data expiration time and
+		// backoff expiration time fields are in the past.
+		now := time.Now()
+		if entry.expiryTime.After(now) || entry.backoffExpiryTime.After(now) {
+			return
+		}
+		evicted = true
+		dc.deleteAndcleanup(key, entry)
+	})
+	return evicted
+}
+
+// resetBackoffState sweeps through the cache and for entries with a backoff
+// state, the backoff timer is cancelled and the backoff state is reset. The
+// return value indicates if any entries were mutated in this fashion.
+//
+// The LB policy invokes this method when the control channel moves from READY
+// to TRANSIENT_FAILURE back to READY. See `monitorConnectivityState` method on
+// the `controlChannel` type for more details.
+func (dc *dataCache) resetBackoffState(newBackoffState *backoffState) bool {
+	if dc.shutdown.HasFired() {
+		return false
+	}
+
+	modified := false
+	dc.keys.iterateAndRun(func(key cacheKey) {
+		entry, ok := dc.entries[key]
+		if !ok {
+			// This should never happen.
+			dc.logger.Errorf("cacheKey %+v not found in the cache while attempting to perform periodic cleanup of expired entries", key)
+			return
+		}
+
+		if entry.backoffState == nil {
+			return
+		}
+		if entry.backoffState.timer != nil {
+			entry.backoffState.timer.Stop()
+			entry.backoffState.timer = nil
+		}
+		entry.backoffState = &backoffState{
+			bs:       newBackoffState.bs,
+			callback: newBackoffState.callback,
+		}
+		entry.backoffTime = time.Time{}
+		entry.backoffExpiryTime = time.Time{}
+		modified = true
+	})
+	return modified
+}
+
+// addEntry adds a cache entry for the given key.
+//
+// Return value evicted indicates if other cache entries were evicted to make
+// space for the current entry. Return value ok indicates if the current entry
```

Comment:
I don't think this is right. `resize` returns true only if a backoff timer was canceled, whereas this says it returns true if there were any evictions.

Reply:
Done. Thanks.